Work with messages as bytes

viktordick opened the pull request:
This is based on the discussion in
`https://github.com/zopefoundation/Products.MailHost/issues/30`
and
`https://github.com/zopefoundation/Products.MailHost/pull/32`.

In order to allow messages with 8bit-encoding to be sent using smtplib,
they have to be prepared as bytes. Since any preparation and header
parsing and manipulation is done in `Products.MailHost`, `zope.sendmail`
only needs to store messages internally as bytes and allow them to be
passed as bytes.

I have not yet added new tests for this added functionality, only fixed the
existing ones.

viktordick commented:
I added some tests and the coverage is again at 100%, although sometimes on my system line 109 in `maildir.py` is not reached because the `FakeOsModule` has a wildcard entry in `self._stat_mode` that matches all created filenames, so I always get the error that the file already exists.

jugmac00 commented on src/zope/sendmail/delivery.py:
```diff
@@ -103,16 +103,27 @@ def newMessageId(self):
+                message = message.encode('utf-8')
+            header = message.split(b'\n\n')[0]
+
+        if bytes is str:
```
In other places in this package this import is used
```
from zope.sendmail._compat import PY2
```
and then

```
if PY2:
```

viktordick commented:
> Maybe a suggestion: under Python 3, most messages passed on to `zope.sendmail` will not be `bytes` but `str` (because under Python 3, working with bytes is really tedious). Therefore, tests covering this typical use case would be very good. Maybe, those tests are already there (as you told us to have 100 % coverage). Otherwise, add them.

If I understand correctly, the only external API for `zope.sendmail` is `MailDelivery.send()`, is it not? That particular function is tested with both bytes and unicode input, including a non-ASCII character in the unicode case. I could also add tests for non-ASCII UTF-8 encoded bytes and for bytes that are not valid UTF-8 (which should still work, in case I want to send something as crazy as a latin1-encoded mail).

d-maurer commented:
Viktor Dick wrote at 2020-7-20 13:18 -0700:
>> Maybe a suggestion: under Python 3, most messages passed on to `zope.sendmail` will not be `bytes` but `str` (because under Python 3, working with bytes is really tedious). Therefore, tests covering this typical use case would be very good. Maybe, those tests are already there (as you told us to have 100 % coverage). Otherwise, add them.
>
>If I understand correctly, the only external API for `zope.sendmail` is `MailDelivery.send()`, is it not? That particular function is tested with both bytes and unicode input, including a non-ASCII character in the unicode case. I could also add tests for non-ASCII UTF-8 encoded bytes and for bytes that are not valid UTF-8 (which should still work, in case I want to send something as crazy as a latin1-encoded mail).

The 100 % coverage likely means that all necessary cases have been
tested. Then no additional tests would be necessary.
My comment was just a precaution.

d-maurer commented on src/zope/sendmail/delivery.py:
```diff
@@ -103,16 +103,27 @@ def newMessageId(self):
+        else:
+            if not isinstance(message, bytes):
+                message = message.encode('utf-8')
+            header = message.split(b'\n\n')[0]
```
Here is a potential likely minor problem: if the message strictly follows RFC 822, it will use `\r\n` rather than `\n` line separators. The effect would be that the whole message would be classified as header - not indented but likely not serious as well.

viktordick commented on src/zope/sendmail/delivery.py:
```diff
@@ -103,16 +103,31 @@ def newMessageId(self):
+            nli = message.find(b'\n')
+            line_sep = b'\n' if nli < 1 or message[nli - 1] != b'\r' \
+                else b'\r\n'
+            header = message.split(line_sep * 2, 1)[0]
```
Since we are mostly interested in the first double newline, we could simply use something like

    nlipos = message.find(b'\r\n\r\n')
    if nlipos == -1:
        nlipos = message.find(b'\n\n')
    header = message[:nlipos]

I guess? But I am fine with the current code also.

viktordick commented:
@jugmac00 I adjusted the Python 2 check according to your comment. I also added a test with a message that is neither ASCII nor UTF-8 (using latin1) to make sure that a message passed as bytes is actually faithfully conserved as bytes throughout and not implicitly converted using UTF-8 somewhere as a bonus.

@d-maurer If you are happy, I could merge this now.

d-maurer commented:
Viktor Dick wrote at 2020-7-23 12:21 -0700:
>@viktordick commented on this pull request.
>
>
>
>> @@ -103,16 +103,31 @@ def newMessageId(self):
>         return "%s@%s" % (left_part, gethostname())
>
>     def send(self, fromaddr, toaddrs, message):
>-        parser = email.parser.Parser()
>-        msg = parser.parsestr(message)
>-        messageid = msg.get('Message-Id')
>+        # Switch the message to be bytes immediately, any encoding
>+        # peculiarities should be handled before.
>+        if message is None:
>+            header = b''
>+        else:
>+            if not isinstance(message, bytes):
>+                message = message.encode('utf-8')
>+            # determine line separator type (assumes consistency)
>+            nli = message.find(b'\n')
>+            line_sep = b'\n' if nli < 1 or message[nli - 1] != b'\r' \
>+                else b'\r\n'
>+            header = message.split(line_sep * 2, 1)[0]
>
>Since we are mostly interested in the first double newline, we could simply use something like
>
>    nlipos = message.find(b'\r\n\r\n')
>    if nlipos == -1:
>        nlipos = message.find(b'\n\n')
>    header = message[:nlipos]
>
>I guess?

Sure, but likely the message does not contain `\r\n`. To find this
out, you look through the whole message.
To locate the line ending, you look only at a few hundred bytes.

d-maurer commented:
Viktor Dick wrote at 2020-7-23 12:38 -0700:
> ...
>@d-maurer If you are happy, I could merge this now.

Go ahead

d-maurer commented:
Viktor Dick wrote at 2020-7-23 12:21 -0700:
>@viktordick commented on this pull request.
> ...
>Since we are mostly interested in the first double newline, we could simply use something like
>
>    nlipos = message.find(b'\r\n\r\n')
>    if nlipos == -1:
>        nlipos = message.find(b'\n\n')
>    header = message[:nlipos]

But after the position is found, your `message[:pos]` is
more efficient than my `split(.., 1)`.

icemac commented:
This PR is missing an entry in the change log. It should be added before a release is cut.
I could cut the release but I am not sure how to document this PR in the change log.

d-maurer commented:
Michael Howitz wrote at 2020-7-30 23:26 -0700:
>This PR is missing an entry in the change log. It should be added before a release is cut.
>I could cut the release but I am not sure how to document this PR in the change log.

I have added a changelog entry (directly in `master` as the PR was
already merged).

icemac commented:
I just released https://pypi.org/project/zope.sendmail/5.1/.
